backend/src/agent/persistence/worker/metrics_reporter.go

```diff
@@ -126,10 +126,24 @@ func (r MetricsReporter) collectNodeMetricsOrNil(
 	return reportMetricsRequest.GetMetrics(), nil
 }
 
-func (r MetricsReporter) readNodeMetricsJSONOrEmpty(runID string, nodeID string) (string, error) {
+func (r MetricsReporter) readNodeMetricsJSONOrEmpty(runID string, nodeStatus workflowapi.NodeStatus) (string, error) {
+	if nodeStatus.Outputs == nil || nodeStatus.Outputs.Artifacts == nil {
```

Comment:
Thank you so much for the fix! It has been there for quite a while...

A quick question here: since mlpipine-metrics is option,  is it possible that user does specify a mlpipeline-metrics json file, but the workflow node fails to produce one? In other words, nodeStatus.Outputs.Artifacts don't contain "mlpipeline-metrics" but that indicates an error instead of indicating that users don't specify mlpipeline-metrics?  I am not familiar with the workflow.NodeStatus, so I am not sure if the info there is purely determined by the workflow runtime info or is also determined by the workflow static configuration.

Other than that, /lgtm /approve

Reply:
I am not sure too, but remember backend added optional to these artifacts for backwards compatability, So we have no way to clearly know a component should output one.